packages/ethereum-group-pcd/src/EthereumGroupPCD.ts

```diff
@@ -0,0 +1,329 @@
+import {
+  ArgumentTypeName,
+  DisplayOptions,
+  PCD,
+  PCDArgument,
+  PCDPackage,
+  SerializedPCD,
+  StringArgument,
+} from "@pcd/pcd-types";
+import {
+  SemaphoreIdentityPCD,
+  SemaphoreIdentityPCDPackage,
+  SemaphoreIdentityPCDTypeName,
+} from "@pcd/semaphore-identity-pcd";
+import {
+  SemaphoreSignaturePCD,
+  SemaphoreSignaturePCDPackage,
+} from "@pcd/semaphore-signature-pcd";
+import {
+  CircuitPubInput,
+  MembershipProver,
+  MembershipVerifier,
+  ProverConfig,
+  PublicInput,
+} from "@personaelabs/spartan-ecdsa";
+import { ethers } from "ethers";
+import { sha256 } from "js-sha256";
+import JSONBig from "json-bigint";
+import { v4 as uuid } from "uuid";
+import { SemaphoreIdentityCardBody as EthereumGroupCardBody } from "./CardBody";
+
+/**
+ * All signature PCDs are 'namespaced' to this pseudo-random nullifier,
+ * so that they cannot be reused by malicious actors across different
+ * applications.
+ */
+export const STATIC_ETH_PCD_NULLIFIER = generateMessageHash(
+  "hardcoded-nullifier"
+);
+
+/**
+ * Hashes a message to be signed with sha256 and fits it into a baby jub jub field element.
+ * @param signal The initial message.
+ * @returns The outputted hash, fed in as a signal to the Semaphore proof.
+ */
+export function generateMessageHash(signal: string): bigint {
+  // right shift to fit into a field element, which is 254 bits long
+  // shift by 8 ensures we have a 253 bit element
+  return BigInt("0x" + sha256(signal)) >> BigInt(8);
+}
+
+export const EthereumGroupPCDTypeName = "ethereum-group-pcd";
+
+export interface EthereumGroupPCDInitArgs {
+  // TODO: how do we distribute these in-package, so that consumers
+  // of the package don't have to copy-paste these artifacts?
+  // TODO: how do we account for different versions of the same type
+  // of artifact? eg. this one is parameterized by group size. Should
+  // we pre-generate a bunch of artifacts per possible group size?
+  // Should we do code-gen?
+  zkeyFilePath: string;
+  wasmFilePath: string;
+}
+
+// We hardcode the externalNullifer to also be your identityCommitment
+// so that your nullifier for specific groups is not revealed when
+// a SemaphoreSignaturePCD is requested from a consumer application.
+export interface EthereumGroupPCDArgs {
+  identity: PCDArgument<SemaphoreIdentityPCD>;
+  signatureOfIdentityCommitment: StringArgument;
+  merkleProof: StringArgument;
+  groupType: StringArgument;
+}
```

Comment:
i think this PCD is complicated enough such that it deserves some fairly thorough documentation